### PR TITLE
Modify RetryHandler.SendRetryAsync Retry check logic and update RetryHandlerTest 

### PR DESCRIPTION
**Pull request by brcaswell**
supersedes brcaswell/msgraph-sdk-dotnet-core#1

### Changes proposed in this pull request
-  address Retry conditional check logic inconsistency between `SendAsync` and `SendRetryAsync` 
-  fixes issue in feature 2.0 that causes Retry to never go beyond original request and 1 retry (2 requests)
-  Updates/Modifies `RetryHandlerTest` `ShouldRetryBasedOnCustomShouldRetryDelegate` 

### Other links
- relates to issue #190 changes
- relates to PR #242 changes

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/pull/266)

### CODE REVIEW

File: src/Microsoft.Graph.Core/Requests/Middleware/RetryHandler.cs [hunk 1]
```diff
@@ -115,7 +115,7 @@ private async Task<HttpResponseMessage> SendRetryAsync(HttpResponseMessage respo
                 // Call base.SendAsync to send the request
                 response = await base.SendAsync(request, cancellationToken);
 
-                if (!ShouldRetry(response) || !request.IsBuffered() || !RetryOption.ShouldRetry(RetryOption.Delay, retryCount, response))
+                if (retryCount >= RetryOption.MaxRetry || !(request.IsBuffered() && (ShouldRetry(response) || RetryOption.ShouldRetry(RetryOption.Delay, 0, response))))
```
Comment (brcaswell) [inline src/Microsoft.Graph.Core/Requests/Middleware/RetryHandler.cs:118]: I need to correct this, it should be passing `retryCount` as a parameter.

note: apart from `retryCount` check relating to iteration, this check is modified to be the inverse of `SendAsync` retry check logic.

File: tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Middleware/RetryHandlerTests.cs [hunk 2]
```diff
@@ -292,40 +295,45 @@ public async Task ShouldRetrytBasedOnRetryAfter(HttpStatusCode statusCode)
...
+            HttpMessageInvoker httpMessageInvoker = new(retryHandler);
+
             // Act
-            var response = await httpMessageInvoker.SendAsync(httpRequestMessage, new CancellationToken());
+            var response = await httpMessageInvoker.SendAsync(request, new CancellationToken());
 
             // Assert
-            Assert.Same(response, finalResponse);
-            Assert.NotSame(response.RequestMessage, httpRequestMessage);
-            Assert.NotNull(response.RequestMessage.Content);
```
Comment (brcaswell) [inline tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Middleware/RetryHandlerTests.cs:326]: removing these checks, they really only cover the mock behavior of `testHttpMessageHandler` type to begin with. 

neither `retryHandler.SendAsync` nor `retryHandler.SendRetryAsync` have any conditionality or functionality that instantiates or modifies `HttpResponseMessage` nor `HttpResponseMessage.RequestMessage`.

**Review by brcaswell — COMMENTED** [review]
Going to do a commit to handle `retryCount` passing

File: src/Microsoft.Graph.Core/Requests/Middleware/RetryHandler.cs (hunk 1 above)
Comment (brcaswell) [reply src/Microsoft.Graph.Core/Requests/Middleware/RetryHandler.cs]: the `retryCount` check related part actually need to go as well.. it was too impactful, handling the expected `ServiceException` therefore had to be added.

**Review by andrueastman — APPROVED** [review]

**Comment by andrueastman** [comment]
@brcaswell Thanks again for this. This is a great help and is much apprecated.

I will update you here once these changes are available via a nuget package.

**Pull request merged**